ogusa_calibrate/cpshousehold.py

```diff
@@ -0,0 +1,138 @@
+import pandas as pd
+import numpy as np
+import microdf as mdf
+import matplotlib.pyplot as plt
+import statsmodels.api as sm
+lowess = sm.nonparametric.lowess
+
+cps = pd.read_csv('https://github.com/PSLmodels/taxdata/raw/master/data/cps.csv.gz',
+                  usecols=['age_head', 'age_spouse', 'elderly_dependents',
+                           'nu18', 'n1820', 'n21', 's006',
+                           'e00200'  # Wages, salaries, and tips. 
+                           # TODO: Switch to AGI or expanded_income,
+                           # from taxcalc output.
+                           ])
+
+def add65(age_spouse):
+  count = 0
+  if(age_spouse>=65):
+    count += 1
+  return count
+age65 = cps.apply(lambda x: add65(x['age_spouse']), axis=1)
+cps['n65'] = cps['elderly_dependents'] + age65
+
+def add1864(n1820, n21, n65):
+  if(n65>=n21):
+    return 0
+  else:
+    return n1820+n21-n65-1
+cps['n1864'] = cps.apply(lambda x: add1864(x['n1820'],x['n21'],x['n65']),axis=1)
+
+isAge = cps['age_head']==21
+temp = cps[isAge].copy(deep=True)
+mdf.add_weighted_quantiles(temp, 'e00200', 's006')
+cols = temp.columns
+
+cps2 = pd.DataFrame(columns=cols)
+for age in cps.age_head.unique():
+  isAge = cps['age_head']==age
+  temp = cps[isAge].copy(deep=True)
+  mdf.add_weighted_quantiles(temp, 'e00200', 's006')
+  cps2 = pd.concat([cps2,temp], ignore_index=True)
+cps2['income_bin'] = pd.cut(cps2['e00200_percentile_exact'],
+                            [0, 25, 50, 70, 90, 99, 100], right=False)
+
+cps2 = cps2.drop(['e00200_percentile_exact',
+           'e00200_percentile', 'e00200_2percentile', 'e00200_ventile',
+           'e00200_decile', 'e00200_quintile', 'e00200_quartile'],
+          axis=1)
+
+cps2 = cps2.groupby(['income_bin', 'age_head']).apply(
+    lambda x: pd.Series({
+        'nu18': mdf.weighted_mean(x, 'nu18', 's006'),
+        'n1864': mdf.weighted_mean(x, 'n1864', 's006'),
+        'n65': mdf.weighted_mean(x, 'n65', 's006'),
+        }))
+cps2.reset_index(inplace = True)
+cps2[cps2['age_head'].between(20,80)]
+
+smoothed18 = []
```

Comment:
Suggested replacement for the remainder of the script (see [colab](https://colab.research.google.com/drive/1ofG9PWl9CWvOYNCvaWVN6PCpQx3b6IBC?usp=sharing#scrollTo=jXHUjV7BzP6-)):
```
def smooth(x, y, frac=0.4):
    """ Produces LOESS smoothed data.
    """
    return pd.Series(lowess(y, x, frac=frac)[:, 1], index=x)


def smooth_all(data):
    """ Return smoothed versions of nu18, n1864, n65.
    """
    return data.groupby(["income_bin", "age_group"]).apply(
        lambda x: smooth(x.age_head, x.n)
    )

cps_long = cps4.drop(columns="index").melt(["age_head", "income_bin"],
                                           var_name="age_group",
                                           value_name="n")
smoothed_wide = smooth_all(cps_long)
smoothed_long = smoothed_wide.reset_index().melt(["age_group", "income_bin"],
                                                 var_name="age_head",
                                                 value_name="n")

# Stack with raw.
cps_long["smoothed"] = False
smoothed_long["smoothed"] = True
combined_long = pd.concat([cps_long, smoothed_long])

# Add the household head. NB: age_head starts at 20 so no need to do for nu18.
combined_long["add_head"] = (
    # n1864 and head age between 18 and 64.
    ((combined_long.age_group == "n1864") &
     combined_long.age_head.between(18, 64)) | 
    # n65 and head age exceeds 64.
    ((combined_long.age_group == "n65") &
     (combined_long.age_head > 64)))
combined_long.n += combined_long.add_head

def plot(data):
    age_group = data.age_group.iloc[0]
    smoothed = data.smoothed.iloc[0]
    title = "Average number of people aged "
    # TODO: Add folder.
    fname = "cps_" + age_group
    if age_group == "nu18":
        title += "0 to 17"
    elif age_group == "n1864":
        title += "18 to 64"
    else:
        title += "65 or older"
    if smoothed:
        title += " (smoothed)"
        fname += "_smoothed"
    tmp.pivot_table("n", "age_head", "income_bin").plot()
    plt.title(title)
    plt.savefig(fname + ".png")

# Create and export all plots.
combined_long.groupby(["age_group", "smoothed"]).apply(plot)
```
I'd stack the PSID data with this too and then add `data_source` as a groupby everywhere to minimize the code. Then just export `combined_long` to a csv.